llrt_modules/src/modules/path.rs

```diff
@@ -1,267 +1,292 @@
+        .map(|n| n.to_str().unwrap())
+        .unwrap_or_default();
+
+    let (name, extension) = name_extname(filename);
+
+    let root = path
+        .components()
+        .next()
+        .and_then(|c| match c {
+            Component::Prefix(prefix) => prefix.as_os_str().to_str(),
+            Component::RootDir => c.as_os_str().to_str(),
+            _ => Some(""),
+        })
+        .unwrap_or_default();
+
+    obj.set("root", root)?;
+    obj.set("dir", parent)?;
+    obj.set("base", [name, extension].concat())?;
+    obj.set("ext", extension)?;
+    obj.set("name", name)?;
+
+    Ok(obj)
+}
+
+fn join(parts: Rest<String>) -> String {
+    join_path(parts.0.iter())
+}
+
+pub fn join_path<S, I>(parts: I) -> String
+where
+    S: AsRef<str>,
+    I: Iterator<Item = S>,
+{
+    let mut empty = true;
+    let result = parts.fold(PathBuf::new(), |mut acc, part| {
+        let part = part.as_ref();
+        if part.starts_with(SEP_PAT) && empty {
+            acc.push(MAIN_SEPARATOR_STR);
+            empty = false;
+        }
+        for sub_part in part.split(SEP_PAT) {
+            if !sub_part.is_empty() {
+                if sub_part.starts_with("..") {
+                    empty = false;
+                    acc.pop();
+                } else {
+                    let sub_part = sub_part.strip_prefix('.').unwrap_or(sub_part);
+                    acc.push(sub_part);
+                    empty = false;
+                    #[cfg(windows)]
+                    if sub_part.ends_with(":") {
+                        acc.push(MAIN_SEPARATOR_STR);
+                    }
+                }
+            }
+        }
+        acc
+    });
+
+    remove_trailing_slash(result)
+}
+
+fn remove_trailing_slash(result: PathBuf) -> String {
+    let path = result.to_string_lossy().to_string();
+    path.strip_suffix(SEP_PAT).unwrap_or(&path).to_string()
+}
+
+fn resolve(path: Rest<String>) -> String {
+    resolve_path(path.iter())
+}
+
+pub fn resolve_path<S, I>(parts: I) -> String
+where
+    S: AsRef<str>,
+    I: Iterator<Item = S>,
+{
+    let result = parts.fold(std::env::current_dir().unwrap(), |mut acc, part| {
+        let part = part.as_ref();
+        if is_absolute(part.into()) {
+            acc = PathBuf::from(part);
+        } else {
+            for sub_part in part.split(SEP_PAT) {
+                if sub_part.starts_with("..") {
+                    acc.pop();
+                } else if sub_part == "." {
+                    // skip
+                } else if sub_part.starts_with("./") {
+                    acc.push(sub_part.strip_prefix("./").unwrap_or(sub_part))
+                } else if sub_part.starts_with(".\\") {
+                    acc.push(sub_part.strip_prefix(".\\").unwrap_or(sub_part))
+                } else {
+                    acc.push(sub_part)
+                }
+            }
+        }
+        acc
+    });
+
+    #[cfg(windows)]
+    return remove_trailing_slash(result).replace("\\", "/");

```

Comment:
If you don't replace ```\```, it will crash on the win platform

```
Assertion failed: m->status == JS_MODULE_STATUS_UNLINKED || m->status == JS_MODULE_STATUS_LINKED || m->status == JS_MODULE_STATUS_EVALUATING_ASYNC || m->status == JS_MODULE_STATUS_EVALUATED, file C:\wt\llrt\target\debug\build\rquickjs-sys-2b4a411d4a450043\out\quickjs.c, line 28585
error: process didn't exit successfully: `target\debug\llrt.exe test -d bundle/js/__tests__/unit` (exit code: 3)
make: *** [Makefile:181: test] Error 3

```

Reply:
Since we iterate trough each char, we should do it in the iteration and not replace afterwards

Reply:
On the win platform, PathBuf uses the \ separator by default.

```rs
    let mut p = PathBuf::new();

    p.push("a");
    p.push("b");

    // a\b
    println!("{}", p.to_string_lossy().to_string());
```

Reply:
Right, I did not consider that. I will merge and add some improvements later. The best approach here would probably be to store an array of string slices rather than a PathBuf and join by slash. But that seems out of scope for this PR.